src/main/java/com/aws/greengrass/mqttclient/IotCoreTopicValidator.java

```diff
@@ -0,0 +1,159 @@
+        if (isMQTT5 && Pattern.matches(SHARED_SUBSCRIPTION_TEMPLATE, topic)) {
+            if (operation == Operation.SUBSCRIBE) {
+                String effectiveTopic = removePrefix(topic, SHARED_SUBSCRIPTION_PREFIX)
+                        .orElseThrow(() -> new MqttRequestException(
+                                "Effective shared subscription topic (without '$share/share-group/' prefix) is empty"));
+                validateEffectiveTopic(effectiveTopic, operation);
+            } else {
+                validateEffectiveTopic(topic, operation);
+            }
+            return;
+        }
+
+        // direct ingest topics
+        if (Pattern.matches(DIRECT_INGEST_TEMPLATE, topic)) {
+            String effectiveTopic = removePrefix(topic, DIRECT_INGEST_PREFIX)
+                    .orElseThrow(() -> new MqttRequestException(
+                            "Effective direct ingest topic (without '$aws/rules/rule-name/' prefix) is empty"));
+            validateEffectiveTopic(effectiveTopic, operation);
+            return;
+        }
+
+        // unknown reserved topic
+        if (isMQTT5) {
+            // rely on IoT Core to perform topic size and forward slash validation,
+            // rather than attempt to keep track of every known IoT reserved topic.
+            // just check a reasonably large topic size to limit large payloads from being sent.
+            if (topic.length() > MAX_LENGTH_FOR_UNKNOWN_RESERVED_TOPIC) {
+                String msg = String.format(
+                        "Reserved topic (%s...) total length is greater than %d bytes of UTF-8 encoded characters "
+                                + "and is most likely over the IoT Core limit of %d bytes (excluding prefixes).",
+                        prefix(topic, 3), MAX_LENGTH_FOR_UNKNOWN_RESERVED_TOPIC, MAX_LENGTH_OF_TOPIC);
+                throw new MqttRequestException(msg);
+            }
+        } else {
+            // treat as normal topic for mqtt3
+            validateEffectiveTopic(topic, operation);
+        }
+    }
+
+    @SuppressWarnings("PMD.UseStringBufferForStringAppends")
+    private static void validateEffectiveTopic(String effectiveTopic, Operation operation) throws MqttRequestException {
+        if (effectiveTopic.chars().filter(num -> num == '/').count() > TOPIC_MAX_NUMBER_OF_FORWARD_SLASHES) {
+            String msg = String.format("The request topic must have no more than %d forward slashes (/)",
+                    TOPIC_MAX_NUMBER_OF_FORWARD_SLASHES);
+            throw new MqttRequestException(msg);
+        }
+        if (effectiveTopic.length() > MAX_LENGTH_OF_TOPIC) {
+            String msg = String.format("The topic size of request must be no "
+                            + "larger than %d bytes of UTF-8 encoded characters. This excludes the first "
+                            + "3 mandatory segments for Basic Ingest topics ($AWS/rules/rule-name/)",
+                    MAX_LENGTH_OF_TOPIC);
```

Comment:
nit: these string errors can be constants. You're always formatting in the same data each time.

Reply:
fixed